Sources/NIOEmbedded/AsyncTestingChannel.swift

```diff
@@ -0,0 +1,521 @@
+    /// and then travelled the `ChannelPipeline` all the way too the back. For data to hit the inbound buffer, the
+    /// last `ChannelHandler` must have send the event either explicitly (by calling
+    /// `ChannelHandlerContext.fireChannelRead`) or implicitly by not implementing `channelRead`.
+    ///
+    /// - note: `NIOAsyncTestingChannel.writeInbound` will fire data through the `ChannelPipeline` using `fireChannelRead`.
+    @inlinable
+    public func readInbound<T>(as type: T.Type = T.self) async throws -> T? {
+        try await self.testingEventLoop.executeInContext {
+            try self._readFromBuffer(buffer: &self.channelcore.inboundBuffer)
+        }
+    }
+
+    /// Sends an inbound `channelRead` event followed by a `channelReadComplete` event through the `ChannelPipeline`.
+    ///
+    /// The immediate effect being that the first `ChannelInboundHandler` will get its `channelRead` method called
+    /// with the data you provide.
+    ///
+    /// - parameters:
+    ///    - data: The data to fire through the pipeline.
+    /// - returns: The state of the inbound buffer which contains all the events that travelled the `ChannelPipeline`
+    //             all the way.
+    @inlinable
+    @discardableResult public func writeInbound<T>(_ data: T) async throws -> BufferState {
+        try await self.testingEventLoop.executeInContext {
+            self.pipeline.fireChannelRead(NIOAny(data))
+            self.pipeline.fireChannelReadComplete()
+            try self._throwIfErrorCaught()
+            return self.channelcore.inboundBuffer.isEmpty ? .empty : .full(self.channelcore.inboundBuffer)
+        }
+    }
+
+    /// Sends an outbound `writeAndFlush` event through the `ChannelPipeline`.
+    ///
+    /// The immediate effect being that the first `ChannelOutboundHandler` will get its `write` method called
+    /// with the data you provide. Note that the first `ChannelOutboundHandler` in the pipeline is the _last_ handler
+    /// because outbound events travel the pipeline from back to front.
+    ///
+    /// - parameters:
+    ///    - data: The data to fire through the pipeline.
+    /// - returns: The state of the outbound buffer which contains all the events that travelled the `ChannelPipeline`
+    //             all the way.
+    @inlinable
+    @discardableResult public func writeOutbound<T>(_ data: T) async throws -> BufferState {
+        try await self.writeAndFlush(NIOAny(data))
+
+        return try await self.testingEventLoop.executeInContext {
+            return self.channelcore.outboundBuffer.isEmpty ? .empty : .full(self.channelcore.outboundBuffer)
+        }
+    }
+
+    /// This method will throw the error that is stored in the `NIOAsyncTestingChannel` if any.
+    ///
+    /// The `NIOAsyncTestingChannel` will store an error if some error travels the `ChannelPipeline` all the way past its end.
```

Comment:
Can we specify what end of the pipeline?

Reply:
Errors can only go to one end of the pipeline.